File: ui/src/lib/control/hello.ts
/** The plaintext version/capability hellos (deliverable D6) — the only frames
 *  sent in the clear. They carry no capability detail, just offered versions,
 *  the session kind, and (for pairing) the rendezvous nonce. Their exact bytes
 *  become the Noise prologue, so any middle-party edit breaks the handshake
 *  rather than silently downgrading it. Mirror of `jeliya-protocol::hello`. */

import { ProtoError, Reader, Writer, bytesEqual } from './codec';
import { Frame, FrameType } from './frame';

/** The 4-byte magic prefixing both hellos: `JCTL`. */
export const MAGIC = new Uint8Array([0x4a, 0x43, 0x54, 0x4c]);

/** The maximum number of versions a client may offer. */
export const MAX_VERSIONS = 8;

/** The all-zero pairing nonce a control (already-paired) session must send. */
export const ZERO_NONCE = new Uint8Array(16);

/** Why a session is being opened. */
export enum SessionKind {
  /** Enroll a new control key (carries a live rendezvous nonce). */
  Pairing = 1,
  /** Exercise an already-installed control key (nonce field is all-zero). */
  Control = 2,
}

function sessionKindFromTag(tag: number): SessionKind {
  if (tag !== SessionKind.Pairing && tag !== SessionKind.Control) {
    throw new ProtoError('bad_enum', 'session_kind');
  }
  return tag;
}

/** The browser's opening offer. */
export class ClientHello {
  constructor(
    readonly versions: number[],
    readonly sessionKind: SessionKind,
    /** 16-byte rendezvous nonce from the QR/link for a pairing session;
     *  all-zero for a control session (validated on decode). */
    readonly pairingNonce: Uint8Array,
  ) {}

  encodeBody(): Uint8Array {
    if (this.versions.length === 0 || this.versions.length > MAX_VERSIONS) {
      throw new ProtoError('bad_count', 'versions');
    }
    if (this.sessionKind === SessionKind.Control && !bytesEqual(this.pairingNonce, ZERO_NONCE)) {
      throw new ProtoError('bad_enum', 'pairing_nonce');
    }
    if (this.pairingNonce.length !== 16) throw new ProtoError('bad_count', 'pairing_nonce');
    const w = new Writer();
    w.putBytes(MAGIC);
    w.putU8(this.versions.length);
    for (const v of this.versions) w.putU16(v);
    w.putU8(this.sessionKind);
    w.putBytes(this.pairingNonce);
    return w.intoBytes();
  }

  static decodeBody(buf: Uint8Array): ClientHello {
    const r = new Reader(buf);
    if (!bytesEqual(r.readArray(4), MAGIC)) throw new ProtoError('bad_magic');
    const count = r.readU8();
    if (count === 0 || count > MAX_VERSIONS) throw new ProtoError('bad_count', 'versions');
    const versions: number[] = [];
    for (let i = 0; i < count; i++) versions.push(r.readU16());
    const sessionKind = sessionKindFromTag(r.readU8());
    const pairingNonce = r.readArray(16);
    r.finish();
    if (sessionKind === SessionKind.Control && !bytesEqual(pairingNonce, ZERO_NONCE)) {
      throw new ProtoError('bad_enum', 'pairing_nonce');
    }
    return new ClientHello(versions, sessionKind, pairingNonce);
  }

  toFrame(): Frame {
    return new Frame(FrameType.ClientHello, this.encodeBody());
  }
}

/** The companion's version selection and minimum-safe floor. */
export class ServerHello {
  constructor(
    /** The chosen version, or `0` to mean "no compatible version". */
    readonly version: number,
    /** The companion-enforced minimum-safe protocol version. */
    readonly minVersion: number,
  ) {}

  encodeBody(): Uint8Array {
    const w = new Writer();
    w.putBytes(MAGIC);
    w.putU16(this.version);
    w.putU16(this.minVersion);
    return w.intoBytes();
  }

  static decodeBody(buf: Uint8Array): ServerHello {
    const r = new Reader(buf);
    if (!bytesEqual(r.readArray(4), MAGIC)) throw new ProtoError('bad_magic');
    const version = r.readU16();
    const minVersion = r.readU16();
    r.finish();
    return new ServerHello(version, minVersion);
  }

  toFrame(): Frame {
    return new Frame(FrameType.ServerHello, this.encodeBody());
  }

  /** Whether the companion found a compatible version. */
  isIncompatible(): boolean {
    return this.version === 0;
  }
}

File: crates/jeliya-protocol/src/hello.rs
//! The plaintext version/capability hellos (deliverable D6). These are the
//! only frames sent in the clear, and they carry no capability detail — just
//! offered versions, the session kind, and (for pairing) the rendezvous nonce.
//! Their exact bytes become the Noise prologue, so any middle-party edit breaks
//! the handshake rather than silently downgrading it.

use crate::codec::{ProtoError, Reader, Writer};
use crate::frame::{Frame, FrameType};

/// The 4-byte magic prefixing both hellos: `JCTL`.
pub const MAGIC: [u8; 4] = *b"JCTL";

/// The maximum number of versions a client may offer.
pub const MAX_VERSIONS: usize = 8;

/// The all-zero pairing nonce a control (already-paired) session must send.
pub const ZERO_NONCE: [u8; 16] = [0u8; 16];

/// Why a session is being opened.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum SessionKind {
    /// Enroll a new control key (carries a live rendezvous nonce; no scoped RPC
    /// is ever accepted on this kind).
    Pairing,
    /// Exercise an already-installed control key (nonce field is all-zero).
    Control,
}

impl SessionKind {
    #[must_use]
    pub fn tag(self) -> u8 {
        match self {
            SessionKind::Pairing => 1,
            SessionKind::Control => 2,
        }
    }

    pub fn from_tag(tag: u8) -> Result<Self, ProtoError> {
        Ok(match tag {
            1 => SessionKind::Pairing,
            2 => SessionKind::Control,
            _ => return Err(ProtoError::BadEnum("session_kind")),
        })
    }
}

/// The browser's opening offer.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ClientHello {
    /// Supported protocol versions in descending preference; 1..=8 entries.
    pub versions: Vec<u16>,
    pub session_kind: SessionKind,
    /// The rendezvous nonce from the QR/link for a pairing session; all-zero
    /// for a control session (validated on decode).
    pub pairing_nonce: [u8; 16],
}

impl ClientHello {
    pub fn encode_body(&self) -> Result<Vec<u8>, ProtoError> {
        if self.versions.is_empty() || self.versions.len() > MAX_VERSIONS {
            return Err(ProtoError::BadCount("versions"));
        }
        if self.session_kind == SessionKind::Control && self.pairing_nonce != ZERO_NONCE {
            return Err(ProtoError::BadEnum("pairing_nonce"));
        }
        let mut w = Writer::new();
        w.put_bytes(&MAGIC);
        w.put_u8(self.versions.len() as u8);
        for v in &self.versions {
            w.put_u16(*v);
        }
        w.put_u8(self.session_kind.tag());
        w.put_bytes(&self.pairing_nonce);
        Ok(w.into_vec())
    }

    pub fn decode_body(buf: &[u8]) -> Result<Self, ProtoError> {
        let mut r = Reader::new(buf);
        if r.read_array::<4>()? != MAGIC {
            return Err(ProtoError::BadMagic);
        }
        let count = r.read_u8()? as usize;
        if count == 0 || count > MAX_VERSIONS {
            return Err(ProtoError::BadCount("versions"));
        }
        let mut versions = Vec::with_capacity(count);
        for _ in 0..count {
            versions.push(r.read_u16()?);
        }
        let session_kind = SessionKind::from_tag(r.read_u8()?)?;
        let pairing_nonce = r.read_array::<16>()?;
        r.finish()?;
        if session_kind == SessionKind::Control && pairing_nonce != ZERO_NONCE {
            return Err(ProtoError::BadEnum("pairing_nonce"));
        }
        Ok(Self {
            versions,
            session_kind,
            pairing_nonce,
        })
    }

    pub fn to_frame(&self) -> Result<Frame, ProtoError> {
        Ok(Frame::new(FrameType::ClientHello, self.encode_body()?))
    }
}

/// The companion's version selection and minimum-safe floor.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ServerHello {
    /// The chosen version, or `0` to mean "no compatible version" (the browser
    /// renders an upgrade prompt; the companion then closes).
    pub version: u16,
    /// The companion-enforced minimum-safe protocol version.
    pub min_version: u16,
}

impl ServerHello {
    #[must_use]
    pub fn encode_body(&self) -> Vec<u8> {
        let mut w = Writer::new();
        w.put_bytes(&MAGIC);
        w.put_u16(self.version);
        w.put_u16(self.min_version);
        w.into_vec()
    }

    pub fn decode_body(buf: &[u8]) -> Result<Self, ProtoError> {
        let mut r = Reader::new(buf);
        if r.read_array::<4>()? != MAGIC {
            return Err(ProtoError::BadMagic);
        }
        let version = r.read_u16()?;
        let min_version = r.read_u16()?;
        r.finish()?;
        Ok(Self {
            version,
            min_version,
        })
    }

    #[must_use]
    pub fn to_frame(&self) -> Frame {
        Frame::new(FrameType::ServerHello, self.encode_body())
    }

    /// Whether the companion found a compatible version.
    #[must_use]
    pub fn is_incompatible(&self) -> bool {
        self.version == 0
    }
}
